x-pack/plugins/stack_connectors/server/connector_types/security/tines/tines.ts

```diff
@@ -0,0 +1,177 @@
+  private urls: {
+    stories: string;
+    agents: string;
+    getRunWebhookURL: (webhook: TinesWebhookObject) => string;
+  };
+
+  constructor(params: ServiceParams<TinesConfig, TinesSecrets>) {
+    super(params);
+
+    this.urls = {
+      stories: `${this.config.url}${API_PATH}/stories`,
+      agents: `${this.config.url}${API_PATH}/agents`,
+      getRunWebhookURL: (webhook) =>
+        `${this.config.url}${WEBHOOK_PATH}/${webhook.path}/${webhook.secret}`,
+    };
+
+    this.registerSubActions();
+  }
+
+  private registerSubActions() {
+    this.registerSubAction({
+      name: SUB_ACTION.STORIES,
+      method: 'getStories',
+      schema: TinesStoriesActionParamsSchema,
+    });
+
+    this.registerSubAction({
+      name: SUB_ACTION.WEBHOOKS,
+      method: 'getWebhooks',
+      schema: TinesWebhooksActionParamsSchema,
+    });
+
+    this.registerSubAction({
+      name: SUB_ACTION.RUN,
+      method: 'runWebhook',
+      schema: TinesRunActionParamsSchema,
+    });
+
+    this.registerSubAction({
+      name: SUB_ACTION.TEST,
+      method: 'runWebhook',
+      schema: TinesRunActionParamsSchema,
+    });
+  }
+
+  private getAuthHeaders() {
+    return { 'x-user-email': this.secrets.email, 'x-user-token': this.secrets.token };
+  }
+
+  private async tinesApiRequest<R extends TinesBaseApiResponse, T>(
+    req: SubActionRequestParams<R>,
+    reducer: (response: R) => T
+  ): Promise<T & { incompleteResponse: boolean }> {
+    const response = await this.request<R>({
+      ...req,
+      params: { ...req.params, per_page: API_RESULTS_PER_PAGE },
+    });
+    return {
+      ...reducer(response.data),
+      incompleteResponse: response.data.meta.pages > 1,
+    };
+  }
+
+  protected getResponseErrorMessage(error: AxiosError): string {
+    if (!error.response?.status) {
+      return 'Unknown API Error';
+    }
+    if (error.response.status === 401) {
+      return 'Unauthorized API Error';
+    }
+    return `API Error: (${error.response?.status}) ${error.response?.statusText}`;
+  }
+
+  public async getStories(): Promise<TinesStoriesActionResponse> {
+    return this.tinesApiRequest(
+      {
+        url: this.urls.stories,
+        headers: this.getAuthHeaders(),
+        responseSchema: TinesStoriesApiResponseSchema,
+      },
+      storiesReducer
+    );
+  }
+
+  public async getWebhooks({
+    storyId,
+  }: TinesWebhooksActionParams): Promise<TinesWebhooksActionResponse> {
+    return this.tinesApiRequest(
+      {
+        url: this.urls.agents,
+        params: { story_id: storyId },
+        headers: this.getAuthHeaders(),
+        responseSchema: TinesWebhooksApiResponseSchema,
+      },
+      webhooksReducer
+    );
+  }
+
+  public async runWebhook({
+    webhook,
+    webhookUrl,
+    body,
+  }: TinesRunActionParams): Promise<TinesRunActionResponse> {
+    if (!webhook && !webhookUrl) {
+      throw Error('Invalid subActionsParams: [webhook] or [webhookUrl] expected but got none');
```

Comment:
This error was needed because of a limitation of the schema validation. This action parameter needs an object with: 
```
body & (webhook | webhookUrl)
```
I haven't found a way to define an object with "one of two" properties, so I added a _maybe_ on both and created this error. The schema validation errors are not translated, so I did not translate this one, I followed the same message format.
This error can not happen under normal conditions, however, a custom API request with only the body can make it to this point, so I added this error only to prevent the request be sent with no url.